python/rascaline/power_spectrum.py

```diff
@@ -0,0 +1,245 @@
+                else:
+                    spx_1 = block_1
+                    spx_2 = block_2
+
+                # Maybe we have to distinguis between if
+                factor = sqrt(2) / sqrt(2 * ell + 1)
+
+                properties = Labels(
+                    names=["n1", "n2"],
+                    values=np.array(
+                        [
+                            properties_1.tolist() + properties_2.tolist()
+                            for properties_1 in spx_1.properties
+                            for properties_2 in spx_2.properties
+                        ],
+                        dtype=np.int32,
+                    ),
+                )
+
+                # Compute the invariants by summation and store the results this is
+                # equivalent to an einsum with: ima, imb -> iab
+                data = factor * np.matmul(spx_1.values.swapaxes(1, 2), spx_2.values)
+
+                new_block = TensorBlock(
+                    values=data.reshape(data.shape[0], -1),
+                    samples=spx_1.samples,
+                    components=[],
+                    properties=properties,
+                )
+
+                for parameter in spx_1.gradients_list():
+                    if parameter == "positions":
+                        _positions_gradients(new_block, spx_1, spx_2, factor)
+
+                keys.append((ell, species_center))
+                blocks.append(new_block)
+
+        keys = Labels(
+            names=["l", "species_center"],
+            values=np.array(keys, dtype=np.int32),
+        )
+
+        return TensorMap(keys, blocks).keys_to_properties("l")
+
+
+def _positions_gradients(new_block, block_1, block_2, factor):
+    gradient_1 = block_1.gradient("positions")
+    gradient_2 = block_2.gradient("positions")
+
+    if len(gradient_1.samples) == 0 or len(gradient_2.samples) == 0:
+        gradients_samples = Labels.empty()
+        gradient_values = np.array([]).reshape(0, 1, len(new_block.properties))
+    else:
+        # The "sample" dimension in the power spectrum gradient samples do
+        # not necessarily matches the "sample" dimension in the spherical
+        # expansion gradient samples. We create new samples by creating a
+        # union between the two gradient samples.
+        gradients_samples_values = np.unique(
+            np.concatenate([gradient_1.samples, gradient_2.samples])
+        )
+
+        gradients_samples = Labels(
+            names=gradient_1.samples.names,
+            values=np.asarray(gradients_samples_values.tolist()),
+        )
+
+        gradient_values = np.zeros(
+            [gradients_samples.shape[0], 3, len(new_block.properties)]
+        )
+
+        # this is equivalent to an einsum with: ixma, imb -> ixab
+        gradient_1_values = factor * np.matmul(
+            gradient_1.values.swapaxes(2, 3),
+            block_2.values[gradient_1.samples["sample"], np.newaxis, :, :],
```

Comment:
that does not quite work, if the two blocks have different central species they will have different value samples, and we can not index block_2 with the samples of 1